Features/task 2

JiriLojda opened the pull request.

Commit:
KA-86 Create ItemList.DatabaseLayer project

Commit:
KA-86 Move Item to new project Contracts

Commit:
KA-86 Create IItemsRepository

Commit:
KA-86 Create ItemsRepository with dummy methods

Commit:
KA-86 Make ItemsRepository methods async

Commit:
KA-86 Add dependency injection

Install Unity
Create IIoCContainer interface describing RegisterRequestScope
Create UnityAdapter as Adapter between IIoCContainer and UnityContainer
Create IoCBootstraper project with Bootstraper
Create RegisterTypes method in Bootstraper calling RegisterTypes from Bootstraper of DatabaseLayer
Configure Unity in WebApiConfig's Register method

Commit:
KA-86 Refactor IItemsRepository, ItemsRepository so methods take correct arguments

Commit:
KA-86 Refactor controler to use repository, update tests

Commit:
KA-86 Create IGuidGenerator describing GenerateGuid method

Commit:
KA-86 Create GuidGenerator in new ServiceLayer project

Commit:
KA-86 Register GuidGenerator with IGuidGenerator for IoC

Commit:
KA-86 Add IGuidGenerator dependency into constructor of controller

Update tests to inject GuidGenerator

Commit:
KA-86 Remove unnecessary usings in all files

Commit:
KA-86 Fix incorrect reference check in Put test

PetrSvirak commented on ItemList/AppBootstrapers/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("AppBootstrapers")]
```
And `AssemblyInfo.cs` is the 3ʳᵈ place (1ˢᵗ is filesystem path, 2ⁿᵈ are `csproj`s and `sln` files) one needs to fix when renaming a project.

PetrSvirak commented on ItemList/AppBootstrapers/ItemList.IocBootstraper.csproj:
```diff
@@ -0,0 +1,95 @@
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>AppBootstrapers</RootNamespace>
+    <AssemblyName>AppBootstrapers</AssemblyName>
```
Your assembly name and root namespace and project name and project folder name should be same.

PetrSvirak commented on ItemList/AppBootstrapers/UnityBootstraper.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public static void RegisterTypes(IIoCContainer container)
+        {
+            container.RegisterType<IItemsRepository, ItemsRepository>(new HierarchicalLifetimeManager());
```
If this would not require either internal visibility or publicity of contracts' implementations, it would be actually best solution as the consumer of the contract would decide how it will use the implementation. On the other hand, it takes away some ability of the contractor (~ implementation) to decide what way will it implement a contract – e.g. repository could be implemented as a singleton that exists throughout all application and cares about opening and closing connections and stuff internally; then forcing multiple instances could break database consistency for example. (of course, the scenario I described contains some errors in judgment, however, design is more about being error-prone or error-proof than expecting people not to make errors)

PetrSvirak commented on ItemList/AppBootstrapers/ItemList.IocBootstraper.csproj:
```diff
@@ -0,0 +1,95 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
`AppBootstrapers` folder should not be here at all, should it? :wink:

PetrSvirak commented on ItemList/ItemList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -54,6 +63,8 @@ public void SetUp()
             var id = new Guid("331c43f5-11af-43a4-83d1-7d949ae5a8d7");
             var expectedItem = new Item { Id = id, Value = "text3" };
 
+            _repositoryMock.Get(Arg.Is(id)).Returns(Task.FromResult(expectedItem));
```
Mockery (pun intended) is part of `arrange` block of the test. You should always have exactly 3 empty-line-separated blocks in each test method. These blocks are called AAA (Arrange-Act-Assert). Anytime you feel a block is too long (or reused in multiple tests, etc.), consider extracting it or its parts into standalone helper methods withing the fixture.

PetrSvirak commented on ItemList/ItemList.Api/App_Start/WebApiConfig.cs:
```diff
@@ -10,6 +10,11 @@ public static class WebApiConfig
+            config.DependencyResolver = new UnityResolver(container);
+
+            var unityAdapter = new UnityAdapter(container);
+            Bootstraper.RegisterTypes(unityAdapter);
```
The reason behind extracting a bootstrapper (mind you noticed the double `p` 😉 ) was streamline references/dependencies in your `Api` project. By placing this code here, you succeeded only partially. You should also hide concrete container implementing the dependency injection (i.e. `Unity`) and allow easy swapping of such containers (such as with `CastleWindows` or `QuickInject` or …) through this abstraction. 

You mustn't reference `Unity` in your `Api` at all. All this code should be placed in a `DependencyInjection` assembly. While scaffolded comment suggests otherwise, this is not a right place to register your adapter. It should have been in standalone static class with its own static `Register` method – and this class is actually only class that should be public in your `DepencyInjection` assembly and this `Register` method is actually the only DI-related method that should be called (directly) in `global.asax.cs`.

PetrSvirak commented on ItemList/ItemList.Api/DependecyInjection/UnityAdapter.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace ItemList.Api.DependecyInjection
+{
+    public class UnityAdapter : IIoCContainer
```
Again, this should be an internal implementation of the contract in your `DependencyInjection` assembly.

PetrSvirak commented on ItemList/ItemList.Api/UnityResolver.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace ItemList.Api
+{
+    public class UnityResolver : IDependencyResolver
```
You see here, now you are making your `Api` dependant on `Unity`. Would it not make more sense to make your `DependencyInjection` assembly dependant on `System.Web.Http` since you write a web application, rather than vice versa?

PetrSvirak commented on ItemList/ItemList.Api/UnityResolver.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public void Dispose()
+        {
+            Container.Dispose();
```
Do not be afraid of expression bodied function, especially for one-liners like these.

PetrSvirak commented on ItemList/ItemList.Contracts/DatabaseLayer/IItemsRepository.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public interface IItemsRepository
+    {
+        Task<IEnumerable<Item>> GetAll();
```
Preferably put empty lines between members of same kind (`methods` in this case).

PetrSvirak commented on ItemList/ItemList.DatabaseLayer/ItemList.DatabaseLayer.csproj:
```diff
@@ -0,0 +1,61 @@
+    <ProjectGuid>{85498898-533E-4446-B3D2-5E51E05635C0}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>ItemList.DatabaseLayer</RootNamespace>
```
Remove the `Layer` suffix (do not forget to remove all 3 aspects of the project; this applies to `db` as well)

PetrSvirak commented on ItemList/ItemList.IoCBootstraper/ItemList.IoCBootstraper.csproj:
```diff
@@ -0,0 +1,68 @@
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>ItemList.IoCBootstraper</RootNamespace>
+    <AssemblyName>ItemList.IoCBootstraper</AssemblyName>
```
The name `IoCBootstraper` is, even without the missing double `p` and with abbreviations, as long as the Cinderella story and almost unpronounceable. What about using name `.Api.Bootstrap`?

PetrSvirak commented on ItemList/ItemList.DatabaseLayer/Bootstraper.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public static class Bootstraper
+    {
+        public static void RegisterTypes(IIoCContainer container)
```
[Static class with a static method](http://cdn.pophangover.com/wp-content/uploads/2012/07/disgusted-oh-god-why-text.png)? 😉 

This makes every developer remember too many things when implementing new assembly (what is the name of the class I should create? What method(s) should I put in? What was the interface with DI container abstraction?).

It is much easier to implement an `IBootsrapper` interface that forces you to implement this method. Also, if you were using `AutoMapper` you would like to have another method for registering mappings in this interface as well – whether a developer would implement DI or mapping they would be forced to think about the other usual part of bootstrap.

PetrSvirak commented on ItemList/ItemList.IoCBootstraper/Bootstraper.cs:
```diff
@@ -0,0 +1,13 @@
+        public static void RegisterTypes(IIoCContainer container)
+        {
+            DatabaseLayer.Bootstraper.RegisterTypes(container);
+            ServiceLayer.Bootstraper.RegisterTypes(container);
```
That is not that bad. Not bad at all. However, only important part of this lines is the name of the assembly. While there is no way in .NET one can pass strongly-typed assemblies around, it is possible to use generics and be more expressive on what you are doing here.

Read about [builder pattern](https://sourcemaking.com/design_patterns/builder), now (with the suggestion [from static bootstrapper class](https://github.com/davidcechak/kentico-onboarding-cs/pull/3/files#diff-abf01a48c66cbef862963fb15689f785R9)) you want to build a collection of all required bootstrappers and at one point call something like `RegisterAll`.

Another point of view could be that you want to provide a fluent interface to add new bootstrappers, so one types as few lines of code as possible needing to know, well, not much about DI structure.

In the comment mentioned above, I actually wrote about `AutoMapper` – if this was the case, you would actually hold acknowledgment of these assemblies' bootstrappers at one place and just called it once for DI registration and once for mapper registration (that would also be part of DI container initialization, btw).

PetrSvirak commented on ItemList/ItemList.ServiceLayer/Utils/GuidGenerator.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using ItemList.Contracts.ServiceLayer;
+
+namespace ItemList.ServiceLayer.Utils
```
It would be better called `Utilities` 😉 But, the name is too general, anyway. This folder might contain `Wrappers` (in the meaning of wrappers of static classes) or `Identifiers` (meaning you expect more ID type to be generated here), …

PetrSvirak commented on ItemList/ItemList.ServiceLayer/Utils/GuidGenerator.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public Guid GenerateGuid()
+        {
+            return new Guid();
```
:rofl:

PetrSvirak replied:
OK, that was too savage. You should have written tests for the service ;-)

PetrSvirak commented on ItemList/ItemList.Contracts/Models/Item.cs:
```diff
@@ -1,7 +1,6 @@
-using System.Collections.Generic;
 
-namespace ItemList.Api.Models
+namespace ItemList.Contracts.Models
```
I see some empty lines down the way…

PetrSvirak commented on ItemList/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -1,48 +1,52 @@
+        private readonly IGuidGenerator _guidGenerator;
 
-        public ItemsController()
+        public ItemsController(IItemsRepository repository, IGuidGenerator guidGenerator)
```
You now have dependency injection, so this constructor should not be a thing anymore. Use your container abstraction to register an IItemUrlHelper implementation. You will have to figure out what to do with circular reference. Either move assembly bootstrappers registration to this project or move the helper to the bootstrapper project (which does not make much sense, really) or create a new project with reference to `System.Web` (e.g. `.Api.Services` or `.Services.Web`) or add a reference to `System.Web` to existing service layer (actually do not even think about this one 😇, Susan can surely tell you why) – these are quite all the options you have here; discuss what would be best in your opinion and feel free come with any other I might have overseen.

PetrSvirak commented on ItemList/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -1,48 +1,52 @@
         public async Task<IHttpActionResult> Post(Item item)
         {
-            item = await Task.FromResult(new Item { Id = new Guid("5081544A-5584-4449-B0CD-72B2BFF0AF30"), Ueid = "Hello Susan", Value = "text4" });
+            item.Id = _guidGenerator.GenerateGuid();
```
I do not like the fact that you transform the `Item` within the controller. This should be done in a service layer (since you have one anyway) and the `guidGenerator` should be a service's dependency. You should also not modify obtained `item`, but rather create a new one with copied properties of your interest (this is a motivation that would normally lead to the AutoMapper and separation of application data model and view model). Also, consider where should the `_repository.Create(item)` take place.

PetrSvirak commented on ItemList/ItemList.IoCBootstraper/Bootstraper.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace ItemList.IoCBootstraper
+{
+    public static class Bootstraper
```
I would love to distinguish this manager of bootstrappers or the master bootstrapper from others. What about calling the class simply _Bootstrap_?

PetrSvirak commented on ItemList/ItemList.ServiceLayer/Utils/GuidGenerator.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace ItemList.ServiceLayer.Utils
+{
+    public class GuidGenerator : IGuidGenerator
```
I don't like type name being mentioned here. Maybe if another suffix than `generator` was used. Also, the problem with `generator` is that a class bearing that name should somewhat assemble the result.

You do not intend to do that (care about uniqueness and the way individual `GUID`s are created). Your intention seems to be a bit different - you actually want to be provided with an ID on demand (this is what a `service` does) and you do not care much about its type or internal structure (as far as it fits with `Item`, you are OK with whatever you get). Something along `IdentifierService` might be a bit more convenient, but I'm open to your other suggestions here.

Also, consider the scenario where you had a single DB table somewhere with pre-generated `GUID` in it. This class of yours would give you one anytime you ask, would you still call it a `generator`?

PetrSvirak reviewed (changes requested):
I quite like what I see here, guys. There are just some conceptual things that need to be sorted out.

I particularly miss tests for your service layer (actually, you should too!)

Make a solution and underlying file system structure more readable by using `Source` and `Tests` folder.

Once you read all the comments, get back to this one, please: Draw your very structure of the application now and add the `.DependencyInjection` and `.Api.Services` assemblies to it. Now, consider if you want to live with as many assemblies as you have now and visualize what would squish individual assemblies look like/caused. Then discuss with me or Susan what would be the best approach for you and why (there is no "correct" solution here, that's you need a duck or its enhanced chatty version in any instance you choose (~ me or Susan)).

There is one more thing I don't understand. @JiriLojda has his JS client almost completely connected to this API, yet it does not produce correct `JSON` data. It actually produces XML, finally do something with it… :wink:

There is quite some renaming based on this review, so a note or two on preserving the history and renaming effectively: 
* When renaming or moving project(s), do not use _Visual studio_ – close it actually – it takes ages and does not help much with disk folder structure. Manually (git-history-friendly!) rename project(s) folder(s) and then use notepad++ or something - after all, it's but a bunch of XML files and a pseudo XML file (.sln). You will need to change few lines in `sln` files and then, when renaming, `csproj` of the project in question and all `csproj`s it is referenced in. That's all and quite straight forward – once you try it, you won't get back! :rofl:
* When renaming an interface and implementation, use R# (`CTRL+R,R`) on the interface – it will suggest the renaming of all instance variables and implementations as well and will practically do the job in 3 clicks. Then you will see new `untracked` and `modified` files in `git status` (suggesting delete and add was performed), but once you stage everything together, the command should detect `rename` as expected.
* Isolate all renames to commits – renaming and interface and implementations? One commit. Renaming 2 projects? Two commits. …

Commit:
KA-86 Delete AppBootstraper from file system

Commit:
KA-86 Move source and tests projects to corresponding folders

Move Source projects to Source folder and Tests projects to Tests folder
Edit csprojs and sln accordingly

Commit:
KA-86 Refactor tests to match arrange-act-assert principle

Commit:
KA-86 Rewrite UnityResolver one line method to expression bodied

Commit:
KA-86 Add empty lines between methods declarations in IItemsRepository

Commit:
KA-86 Create IBootstrapper interface

Commit:
KA-86 Implement IBootstrapper in Bootstrapers

Bootstrapers in DatabaseLayer and ServiceLayer implement IBootstrapper interface
Bootstraper in IoCBootstraper use new Bootstrapers from DatabaseLayer ServiceLayer

Commit:
KA-86 Rename IoCBootstraper.Booststraper to IoCBootstraper.Bootstrap

Commit:
KA-86 Move Unity container register and it's adapter to IoCBoostraper

Create static method to register the container and use it in app start to configure DI container
Move UnityResolver to IoCBootstraper

Commit:
KA-86 Register ItemUrlHelper in new Api.Services project

Create Bootstrapper class in Api.Services Registering ItemUrlHelper

Commit:
KA-86 Rename GuidGenerator to IdentifierService, add tests

Rename GenerateGuid to GetIdentifier

Commit:
KA-86 Create ItemStoringService, with tests, and use it in controller

Move ItemEqualityComparer to a new project Contracts.Tests.Base
Make IdentifierService internal

Commit:
KA-86 Refactor Post_ValidItem_ReturnsDummyItemAndUrl

Commit:
KA-86 Rename Bootstraper to Bootstrapper in Database and Service Layer

Commit:
KA-86 Add Async sufix to all async methods

Commit:
KA-86 Rename utils folder in ServiceLayer to Wrappers

Commit:
KA-86 Rename DatabaseLayer to Database and ServiceLayer to Services

Commit:
KA-86 Rename IoCBootstraper to DependencyInjection

Commit:
KA-86 Fix all namespaces in the solution

Commit:
KA-86 Remove unnecessary usings in the solution

Commit:
KA-86 Remove unnecessary empty lines in Item.cs

Commit:
KA-86 Use ItemStoringService in ItemController, update tests

Commit:
KA-86 Add configuration, to use CamelCasePropertyNames

Commit:
KA-86 Create IIncludeBootstraper interface in Contracts

Commit:
KA-86 Create IResolveBootrappersTypes

Commit:
KA-86 Rewrite IoC container registration

Begin IoC registration in Global.asax with BoostrapConfig.Register
BoostrapConfig calls Create(httpConfig) on statit class BootstrapFactory
in BF is initialized property Bootstrapers using BootstrappersBuilder
BB has Include method to add Boostrapers into final list of registered boostrapers
BF.Create creates WebBootstrap passing config received as argument and list of boostrapers
UWB.RegisterDependencies method calls register types on all boostrapers received in constructor
UWB.RegisterDependencies is called in BoostraperConfig

PetrSvirak commented on ItemList/Source/ItemList.Api/App_Start/WebApiConfig.cs:
```diff
@@ -10,20 +8,16 @@ public static class WebApiConfig
-            Bootstraper.RegisterTypes(unityAdapter);
-
+            config.Formatters.JsonFormatter.SerializerSettings.ContractResolver = new CamelCasePropertyNamesContractResolver();
+            config.Formatters.JsonFormatter.UseDataContractJsonSerializer = false;
```
Wrap these formatting into standalone `config` class, there is no reason to mix multiple purposes into single general config (see how `global.asax` clearly states what initialization does without meddling in implementations details while here are details of two clearly distinguishable configurations).

PetrSvirak commented on ItemList/Source/ItemList.Contracts/Api/IIoCContainer.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace ItemList.Contracts.Api
+{
+    public interface IIoCContainer
```
* Move it to the same namespace as `IBootstrap` and `IBootstrapper` – these are actually quite related.
* Renaming to simple `IContainer` or `IDependecyInjectionContainer` or `IInversionOfControlContainer` or something even better (definitely get rid of the `IoC` abbreviation)

PetrSvirak commented on ItemList/Source/ItemList.DependencyInjection/Builder/UnityWebBootstrap.cs:
```diff
@@ -0,0 +1,37 @@
+                instance.RegisterTypes(unityAdapter);
+            }
+
+            container.RegisterInstance<IBootstrap>(this);
```
This line actually deserves a comment – why is it there and what does it mean (do not forget that all contracts should have an implementation and in a bigger system, you would actually want the system not to even start if there is a contract implementation missing).

PetrSvirak commented on ItemList/Source/ItemList.Contracts/Bootstrap/IBootstrap.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace ItemList.Contracts.Bootstrap
```
You actually call it `.DependencyInjection` now. It is somewhat misleading having `.DependencyInjection` implement `.Bootstrap` contract. Consider renaming the namespace.

PetrSvirak commented on ItemList/Tests/ItemList.Api.Tests/ItemsControllerTests.cs:
```diff
@@ -46,9 +46,9 @@ public void SetUp()
                 new Item { Id = new Guid("83aa9154-2b5f-49b7-b7af-25cab7bf2159"), Value = "text2" }
             };
-            _repositoryMock.GetAll().Returns(Task.FromResult(expectedItems));
+            _repositoryMock.GetAllAsync().Returns(Task.FromResult(expectedItems));
```
You should actually rename test as well, should you not?

PetrSvirak commented on ItemList/Tests/ItemList.Services.Tests/ItemStoringServiceTests.cs:
```diff
@@ -0,0 +1,55 @@
+            _identifierServiceMock = Substitute.For<IIdentifierService>();
+
+            _itemStoringService = new ItemStoringService(_identifierServiceMock, _repositoryMock);
+
```
Free line strikes again! :wink:

PetrSvirak commented on ItemList/Tests/ItemList.Services.Tests/ItemStoringServiceTests.cs:
```diff
@@ -0,0 +1,55 @@
+            _identifierServiceMock.GetIdentifier().Returns(expectedItem.Id);
+            _repositoryMock.CreateAsync(Arg.Do<Item>(item => { itemSentToRepository = item; })).Returns(Task.CompletedTask);
+
+
```
...and again :wink:

PetrSvirak reviewed (changes requested):
There are just minor details that need fixing, otherwise, it is all done and pretty nice solution actually. Thanks.

PetrSvirak commented on ItemList/Source/ItemList.DependencyInjection/Builder/BootstrappersBuilder.cs:
```diff
@@ -0,0 +1,38 @@
+using System.Linq;
+using ItemList.Contracts.Bootstrap;
+
+namespace ItemList.DependencyInjection
```
Fix this namespace as it prevents the build of your committed solution :wink:

Commit:
KA-86 Rename Bootstrap folder to DependencyInjection

IIocContainer to IDependencyInjectionContainer

Commit:
KA-86 Add comment into UnityWebBootstrap

Commit:
KA-86 Rename api tests according to api methods

Commit:
KA-86 Fix BootstrappersBuilder namespace

Rename Bootstrap to DependencyInjection

Commit:
KA-86 Remove unnecessary empty lines

Commit:
KA - 86 Create new FormatterConfig and set JsonFormatter there
